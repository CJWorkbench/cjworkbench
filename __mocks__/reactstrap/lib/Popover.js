import React from 'react'

export default class Popover extends React.PureComponent {
  render () {
    const { innerClassName, isOpen } = this.props
    if (!isOpen) return null

    const props = { ...this.props }
    delete props.isOpen
    delete props.innerClassName
    delete props.toggle
    delete props.target
    delete props.boundariesElement

    props.className = innerClassName || null

    return (
      <div className='mock-reactstrap-Popover'>
        <div {...props}/>
      </div>
    )
  }
}
