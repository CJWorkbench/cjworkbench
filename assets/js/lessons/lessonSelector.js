import { createSelector } from 'reselect'
import { StateWithHelpers } from '../lessons/DoneHelpers'
import { matchLessonHighlight } from '../util/LessonHighlight'

function isStepDone(sectionTitle, stepIndex, stateWithHelpers, step) {
  // Canonical example testJs:
  // `return workflow.selectedWfModule.moduleName === 'Add from URL'`
  const fn = new Function('state', 'workflow', '"use strict"; ' + step.testJs)
  // Give our function a name: makes it easy to debug crashes
  Object.defineProperty(fn, 'name', {
    value: `LessonSection "${sectionTitle}" Step ${stepIndex + 1}`,
    writable: false,
  })

  try {
    return fn(stateWithHelpers, stateWithHelpers.workflow)
  } catch (e) {
    console.error(e)
    console.error('The previous error is a bug in this function, called with these arguments:', fn, stateWithHelpers, stateWithHelpers.workflow)
    return false
  }
}

function calculateActiveStep(stateWithHelpers, sections) {
  // Run each testJs function until one returns false
  for (let sectionIndex = 0; sectionIndex < sections.length; sectionIndex++) {
    const section = sections[sectionIndex]
    const steps = section.steps || []
    for (let stepIndex = 0; stepIndex < steps.length; stepIndex++) {
      const step = steps[stepIndex]
      if (!isStepDone(section.title, stepIndex, stateWithHelpers, step)) {
        return {
          activeSectionIndex: sectionIndex,
          activeStepIndex: stepIndex,
          activeStep: step,
        }
      }
    }
  }

  // all steps complete
  return {
    activeSectionIndex: null,
    activeStepIndex: null,
    activeStep: null,
  }
}

/**
 * Return `false` always, so no component gets lesson-highlighted when there
 * is no lesson.
 */
function testLessonHighlightButThereIsNoLesson(test) { return false }

const getWorkflow = ({ workflow }) => workflow
const getTabs = ({ tabs }) => tabs
const getWfModules = ({ wfModules }) => wfModules
const getModules = ({ modules }) => modules
const getSelectedWfModule = ({ selected_wf_module }) => selected_wf_module
const getLessonData = ({ lessonData }) => lessonData || null

/**
 * Returns the Lesson for the given state.
 *
 * Example:
 *
 * ```
 * import lessonSelector from '../lessons/lessonSelector'
 * function mapStateToProps(state, ownProps) {
 *   const { testHighlight } = lessonSelector(state)
 *   return {
 *     isLessonHighlight: testHighlight({ type: 'WfModule', moduleName: ownProps.moduleName })
 *   }
 * }
 * ```
 *
 * The returned object has the following properties:
 *
 * * `activeSectionIndex`: section containing the step we have arrived at
 * * `activeStepIndex`: step within the section that the user needs to complete
 * * `testHighlight`: function that accepts a `LessonHighlightType` and
 *                          returns `true` iff the active step specifies we
 *                          should highlight it.
 *
 * This function never throws. Lessons may have invalid `testJs` values,
 * especially if they index into arrays and the user does something we don't
 * expect. Those errors will be caught and logged with console.error(). An
 * erroring step counts as unfinished: usually, errors mean we haven't imagined
 * a user doing something -- which probably means the user _didn't_ do what we
 * suggested.
 */
const getLesson = createSelector(
  [ getWorkflow, getTabs, getWfModules, getModules, getSelectedWfModule, getLessonData ],
  (workflow, tabs, wfModules, modules, selectedWfModule, lessonData) => {
    if (lessonData === null) {
      return {
        activeSectionIndex: null,
        activeStepIndex: null,
        testHighlight: testLessonHighlightButThereIsNoLesson
      }
    }

    const stateWithHelpers = new StateWithHelpers({
      workflow,
      tabs,
      wfModules,
      modules,
      selected_wf_module: selectedWfModule
    })
    const { activeSectionIndex, activeStepIndex, activeStep } = calculateActiveStep(stateWithHelpers, lessonData.sections)
    const lessonHighlight = activeStep ? activeStep.highlight : []
    const testHighlight = (test) => matchLessonHighlight(lessonHighlight, test)

    return {
      activeSectionIndex,
      activeStepIndex,
      testHighlight
    }
  })
export default getLesson
