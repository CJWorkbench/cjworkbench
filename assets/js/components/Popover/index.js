import Popover from 'reactstrap/lib/Popover'
import PopoverHeader from 'reactstrap/lib/PopoverHeader'
import PopoverBody from 'reactstrap/lib/PopoverBody'

export { Popover, PopoverHeader, PopoverBody }
