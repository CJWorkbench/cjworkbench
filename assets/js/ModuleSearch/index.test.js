/**
 * Testing Stories:
 * -Renders a search bar
 * -Search bar will render suggestions of modules matching input
 * 
 */
jest.mock('../lessons/lessonSelector', () => jest.fn()) // same mock in every test :( ... we'll live

// https://github.com/FezVrasta/popper.js#how-to-use-popperjs-in-jest
jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js')

  return class {
    static placements = PopperJS.placements

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {}
      }
    }
  }
})


import React from 'react'
import ConnectedModuleSearch, { ModuleSearch } from './index'
import Popover from 'reactstrap/lib/Popover'
import { mount, shallow } from 'enzyme'
import { createStore } from 'redux'
import { Provider } from 'react-redux'
import lessonSelector from '../lessons/lessonSelector'

describe('ModuleSearch', () => {
  const modules = {
    4: {
      id: 4,
      name: "Load from Enigma",
      description: 'Enigma description',
      category: "Add data",
      icon: "url",
      isLessonHighlight: true,
    },
    10: {
      id: 10,
      name: "Filter by Text",
      description: 'Text description',
      category: "Filter",
      icon: "filter",
      isLessonHighlight: false,
    }
  }
  const modulesArray = Object.keys(modules).map(id => modules[id])

  let defaultProps
  beforeEach(() => defaultProps = {
    onClickModuleId: jest.fn(),
    onCancel: jest.fn(),
    index: 2,
    modules: modulesArray,
    isLessonHighlight: false,
  })

  describe('most tests', () => {
    let wrapper
    beforeEach(() => wrapper = mount(<ModuleSearch {...defaultProps}/>))
    afterEach(() => wrapper.unmount())

    let searchField
    beforeEach(() => searchField = wrapper.find('input[name="moduleQ"]'))

    it('matches snapshot', () => { 
      expect(wrapper).toMatchSnapshot()
    })

    it('finds all suggestions by default', () => {
      expect(wrapper.text()).toMatch(/Load from Enigma/)
      expect(wrapper.text()).toMatch(/Filter by Text/)
    })

    it('finds a suggestion matching search input', () => { 
      searchField.simulate('change', {target: {value: 'a'}})
      wrapper.update()
      expect(wrapper.text()).toMatch(/Load from Enigma/)
      expect(wrapper.text()).not.toMatch(/Filter by Text/)

      // Search in description
      searchField.simulate('change', {target: {value: 'description'}})
      wrapper.update()
      expect(wrapper.text()).toMatch(/Load from Enigma/)
      expect(wrapper.text()).toMatch(/Filter by Text/)
    })

    it('calls onCancel on form reset (e.g., clicking button.close)', () => { 
      // search field should be empty at start
      wrapper.find('form').simulate('reset')
      expect(wrapper.prop('onCancel')).toHaveBeenCalled()
    });

    it('calls onCancel on pressing Escape', () => {
      searchField.simulate('keyDown', { keyCode: 27 })
      expect(wrapper.prop('onCancel')).toHaveBeenCalled()
    })

    it('calls onClickModuleId on click', () => {
      wrapper.find('li[data-module-name="Load from Enigma"]').simulate('click')
      expect(wrapper.prop('onClickModuleId')).toHaveBeenCalledWith(4)
    })

    it('should lesson-highlight module', () => {
      expect(wrapper.find('li[data-module-name="Load from Enigma"]').hasClass('lesson-highlight')).toBe(true)
      expect(wrapper.find('li[data-module-name="Filter by Text"]').hasClass('lesson-highlight')).toBe(false)
    })

    it('should sort modules in alphabetical order', () => {
      wrapper.unmount()
      const modules = {
        4: {
          id: 4,
          name: "Z",
          description: 'desc',
          category: "Add data",
          icon: "url",
          isLessonHighlight: true,
        },
        1: {
          id: 1,
          name: "Y",
          description: 'desc',
          category: "Filter",
          icon: "filter",
          isLessonHighlight: false,
        },
        6: {
          id: 6,
          name: "H",
          description: 'desc',
          category: "Filter",
          icon: "filter",
          isLessonHighlight: false,
        },
        7: {
          id: 7,
          name: "A",
          description: 'desc',
          category: "Add data",
          icon: "filter",
          isLessonHighlight: false,
        },
        5: {
          id: 5,
          name: "A",
          description: 'desc',
          category: "Filter",
          icon: "filter",
          isLessonHighlight: false,
        }
      }
      const modulesArray = Object.keys(modules).map(id => modules[id])

      let defaultProps = {
        onClickModuleId: jest.fn(),
        onCancel: jest.fn(),
        index: 2,
        modules: modulesArray,
        isLessonHighlight: false
      }

      wrapper = mount(<ModuleSearch {...defaultProps}/>)

      let resultList = wrapper.prop('modules').map(x => x.name)
      let expectedResult = ['A', 'A', 'H', 'Y', 'Z']
      expect(resultList).toEqual(expectedResult)
    })

    it('should show a popover when hovering on a module', () => {
      expect(wrapper.find(Popover)).toHaveLength(0)
      wrapper.find('li[data-module-name="Filter by Text"]').simulate('mouseEnter')
      wrapper.update()
      expect(wrapper.find(Popover)).toHaveLength(1)
    })
  })
    
  it('should highlight search box based on isLessonHighlight', () => {
    const noHighlight = shallow(<ModuleSearch {...defaultProps} isLessonHighlight={false} />)
    expect(noHighlight.hasClass('lesson-highlight')).toBe(false)

    const yesHighlight = shallow(<ModuleSearch {...defaultProps} isLessonHighlight={true} />)
    expect(yesHighlight.hasClass('lesson-highlight')).toBe(true)
  })

  describe('with store', () => {
    let store
    let wrapper
    let nonce = 0

    function highlight(index, moduleName) {
      lessonSelector.mockReturnValue({
        testHighlight: test => {
          if (index === null) return false
          return test.type === 'Module' && (test.name ? (test.name === moduleName) : true) && test.index === index
        }
      })

      // trigger a change
      store.dispatch({ type: 'whatever', payload: ++nonce })
    }

    beforeEach(() => {
      lessonSelector.mockReset()

      // Store just needs to change, to trigger mapStateToProps. We don't care
      // about its value
      store = createStore((_, action) => ({ modules, ...action.payload }), { modules })

      highlight(null)

      wrapper = mount(
        <Provider store={store}>
          <ConnectedModuleSearch {...defaultProps} alwaysRenderSuggestions={true} />
        </Provider>
      )
    })
    afterEach(() => {
      wrapper.unmount()
    })

    it('loads modules', () => {
      expect(wrapper.text()).toMatch(/Load from Enigma/)
    })

    it('highlights the search box', () => {
      highlight(2, null)
      wrapper.update()
      expect(wrapper.find('.module-search').prop('className')).toMatch(/\blesson-highlight\b/)

      highlight(1, null)
      wrapper.update()
      expect(wrapper.find('.module-search').prop('className')).not.toMatch(/\blesson-highlight\b/)
    })
  })
});

