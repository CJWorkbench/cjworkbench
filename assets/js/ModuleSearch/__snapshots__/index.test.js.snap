// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleSearch most tests matches snapshot 1`] = `
<ModuleSearch
  index={2}
  isLessonHighlight={false}
  modules={
    Array [
      Object {
        "category": "Filter",
        "description": "Text description",
        "icon": "filter",
        "id": 10,
        "isLessonHighlight": false,
        "name": "Filter by Text",
      },
      Object {
        "category": "Add data",
        "description": "Enigma description",
        "icon": "url",
        "id": 4,
        "isLessonHighlight": true,
        "name": "Load from Enigma",
      },
    ]
  }
  onCancel={[MockFunction]}
  onClickModuleId={[MockFunction]}
>
  <div
    className="module-search"
  >
    <form
      className="module-search-field"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        name="moduleQ"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Searchâ€¦"
        type="search"
        value=""
      />
      <button
        className="close"
        title="Close Search"
        type="reset"
      >
        <i
          className="icon-close"
        />
      </button>
    </form>
    <ul
      className="module-search-result-groups"
    >
      <ModuleSearchResultGroup
        activeModuleId={null}
        key="Add data"
        modules={
          Array [
            Object {
              "category": "Add data",
              "description": "Enigma description",
              "icon": "url",
              "id": 4,
              "isLessonHighlight": true,
              "name": "Load from Enigma",
            },
          ]
        }
        name="Add data"
        onClickModuleId={[Function]}
        onMouseEnterModuleId={[Function]}
      >
        <li
          className="module-search-result-group"
          data-name="Add data"
        >
          <h4>
            Add data
          </h4>
          <ul
            className="module-search-results"
          >
            <SearchResult
              category="Add data"
              description="Enigma description"
              icon="url"
              id={4}
              isActive={false}
              isLessonHighlight={true}
              key="Load from Enigma"
              name="Load from Enigma"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <li
                className="module-search-result lesson-highlight"
                data-module-name="Load from Enigma"
                id="module-search-result-4"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                <i
                  className="icon-url"
                />
                <span
                  className="name"
                >
                  Load from Enigma
                </span>
              </li>
            </SearchResult>
          </ul>
        </li>
      </ModuleSearchResultGroup>
      <ModuleSearchResultGroup
        activeModuleId={null}
        key="Filter"
        modules={
          Array [
            Object {
              "category": "Filter",
              "description": "Text description",
              "icon": "filter",
              "id": 10,
              "isLessonHighlight": false,
              "name": "Filter by Text",
            },
          ]
        }
        name="Filter"
        onClickModuleId={[Function]}
        onMouseEnterModuleId={[Function]}
      >
        <li
          className="module-search-result-group"
          data-name="Filter"
        >
          <h4>
            Filter
          </h4>
          <ul
            className="module-search-results"
          >
            <SearchResult
              category="Filter"
              description="Text description"
              icon="filter"
              id={10}
              isActive={false}
              isLessonHighlight={false}
              key="Filter by Text"
              name="Filter by Text"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              <li
                className="module-search-result"
                data-module-name="Filter by Text"
                id="module-search-result-10"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                <i
                  className="icon-filter"
                />
                <span
                  className="name"
                >
                  Filter by Text
                </span>
              </li>
            </SearchResult>
          </ul>
        </li>
      </ModuleSearchResultGroup>
    </ul>
  </div>
</ModuleSearch>
`;
